sysl2/sysl/importer/xsd.go

```diff
@@ -0,0 +1,224 @@
+
+	"aqwari.net/xml/xsd"
+	"github.com/sirupsen/logrus"
+)
+
+func LoadXSDText(args OutputData, text string, logger *logrus.Logger) (out string, err error) {
+	xsd.StandardSchema = [][]byte{} // Ignore all the standard schemas,
+	specs, err := xsd.Parse([]byte(text))
+	if err != nil {
+		return "", err
+	}
+
+	types := TypeList{}
+	for _, schema := range specs {
+		schemaTypes := loadSchemaTypes(schema, logger)
+		types.Add(schemaTypes.Items()...)
+	}
+	types.Sort()
+
+	info := SyslInfo{
+		OutputData:  args,
+		Description: "",
+		Title:       "",
+	}
+
+	result := &bytes.Buffer{}
+	w := newWriter(result, logger)
+	if err := w.Write(info, types, ""); err != nil {
+		return "", err
+	}
+
+	return result.String(), nil
+}
+
+func makeNamespacedType(name xml.Name, target Type) Type {
+	return &ImportedBuiltInAlias{
+		name:   fmt.Sprintf("%s:%s", name.Space, name.Local),
+		Target: target,
+	}
+}
+
+func loadSchemaTypes(schema xsd.Schema, logger *logrus.Logger) TypeList {
+	types := TypeList{}
+
+	var xsdToSyslMappings = map[string]string{
+		"NMTOKEN":      StringTypeName,
+		"integer":      "int",
+		"time":         StringTypeName,
+		"boolean":      "bool",
+		StringTypeName: StringTypeName,
+		"date":         "date",
+	}
+	for swaggerName, syslName := range xsdToSyslMappings {
+		from := xml.Name{Space: "http://www.w3.org/2001/XMLSchema", Local: swaggerName}
+		to := &SyslBuiltIn{syslName}
+		types.Add(makeNamespacedType(from, to))
+	}
+
+	keys := make([]xml.Name, 0, len(schema.Types))
+	for key := range schema.Types {
+		keys = append(keys, key)
+	}
+	sort.Slice(keys, func(i, j int) bool {
+		return strings.Compare(
+			fmt.Sprintf("%s:%s", keys[i].Space, keys[i].Local),
+			fmt.Sprintf("%s:%s", keys[j].Space, keys[j].Local)) < 0
+	})
+
+	for _, name := range keys {
+		data := schema.Types[name]
+		logger.Infof("name: %+v\n", name)
+		if name.Local == "_self" {
+			rootType := data.(*xsd.ComplexType)
+			data = rootType.Elements[0].Type
+			t := FindType(data, &types)
+			if t == nil {
+				t = makeType(name, data, &types, logger)
+				if name.Space != "" {
+					types.Add(makeNamespacedType(name, t))
+				}
+			}
+			x := t.(*StandardType)
+			if !contains("~xml_root", x.Attributes) {
+				x.Attributes = append(x.Attributes, "~xml_root")
+			}
+		} else if t := FindType(data, &types); t == nil {
+			types.Add(makeType(name, data, &types, logger))
+		}
+	}
+
+	return types
+}
+
+func FindType(t xsd.Type, knownTypes *TypeList) Type {
+	if res, found := knownTypes.Find(fmt.Sprintf("%s:%s", xsd.XMLName(t).Space, xsd.XMLName(t).Local)); found {
+		return res
+	}
+	if res, found := knownTypes.Find(xsd.XMLName(t).Local); found {
+		return res
+	}
+	return nil
+}
+
+func makeType(_ xml.Name, from xsd.Type, knownTypes *TypeList, logger *logrus.Logger) Type {
+	switch t := from.(type) {
+	case *xsd.ComplexType:
+		return makeComplexType(t, knownTypes, logger)
+	case *xsd.SimpleType:
+		return makeSimpleType(t, knownTypes, logger)
+	case xsd.Builtin:
```

Comment:
Not sure if this is an issue, case xsd.Builtin: is the only one not dereferenced.

Reply:
xsd.Buiiltin is a typedef to int, the others are structs